pkg/internal/server/grpc/snapshotter_secret.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpc
+
+import (
+	"context"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumesnapshot/v1"
+	snapshotutils "github.com/kubernetes-csi/external-snapshotter/v8/pkg/utils"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// getSnapshotterCredentials returns the snapshotter secret associated with VolumeSnapshotClass, if any.
+// The VolumeSnapshotClass is fetched by name if specified, or else the default VolumeSnapshotClass
+// for the CSI driver is used.
+func (s *Server) getSnapshotterCredentials(ctx context.Context, vsi *volSnapshotInfo) (map[string]string, error) {
+	secretRef, err := s.getSnapshotterSecretRef(ctx, vsi)
+	if err != nil {
+		return nil, err
+	}
+
+	if secretRef == nil {
+		return nil, nil
+	}
+
+	// snapshotutils.GetCredentials gets the secrets for the external-snapshotter CSI calls.
+	secretMap, err := snapshotutils.GetCredentials(s.kubeClient(), secretRef)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, msgUnavailableFailedToGetCredentialsFmt, err) // error contains secret ref
+	}
+
+	return secretMap, nil
+}
+
+// getSnapshotterSecretRef returns the name of the snapshotter secret from a VolumeSnapshotClass.
+// The VolumeSnapshotClass is fetched by name if specified, or else the default VolumeSnapshotClass
+// for the CSI driver is used.
+// On successful return, either both secretNs and secretName are set or neither are set.
+func (s *Server) getSnapshotterSecretRef(ctx context.Context, vsi *volSnapshotInfo) (*corev1.SecretReference, error) {
+	var vsClass *snapshotv1.VolumeSnapshotClass
+	var err error
+
+	if vsi.VolumeSnapshot.Spec.VolumeSnapshotClassName != nil {
+		vsClass, err = s.getVolumeSnapshotClass(ctx, *vsi.VolumeSnapshot.Spec.VolumeSnapshotClassName)
+	} else {
+		vsClass, err = s.getDefaultVolumeSnapshotClassForDriver(ctx, vsi.DriverName)
+	}
```

Comment:
I'm handling the default volume snapshot class case here but from what I can see in the [external-snapshotter common controller code](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller.go#L751) the case is not handled!